StudentPlan.py

```diff
@@ -23,50 +32,97 @@ def accommodate(self, course, correq=False):
             min_position = max(min_position, self.accommodate(pre_requisite))
         for co_requisite in course.co_requisites:
             min_position = max(min_position, self.accommodate(co_requisite, correq=True))
-        min_position = self.bops(min_position=min_position, course=course)
+        min_position = self.set_to_earliest_possible_semester(min_position=min_position, course=course)
         return min_position + 1 if not correq else course.position
 
-    def add_semester(self, position):
-        semester = Semester(max_credits=self.max_credits, position=position)
+    def add_semester(self, position, past=False):
+        semester = Semester(max_credits=self.max_credits, position=position, past=past)
         self.semesters.append(semester)
         return semester
 
-    def bops(self, course, min_position):
-        if min_position >= len(self.semesters):
+    def force_accommodate(self, position, course):
+        if not self.get_semester(position):
+            new_semester = self.add_semester(position)
+            new_semester.add_course(course)
+
+    def get_semester(self, position):
+        for semester in self.semesters:
+            if semester.position == position:
+                return semester
+
+    def set_to_earliest_possible_semester(self, course, min_position):
+        if not self.get_semester(min_position):
             self.add_semester(position=min_position)
+        self.semesters = sorted(self.semesters, key=lambda s: s.position,)
         for semester_index in range(min_position, len(self.semesters)):
-            if not self.semesters[semester_index].is_full:
-                if self.semesters[semester_index].credit_hours + course.credit_hours <= self.semesters[semester_index].max_credits:
-                    self.semesters[semester_index].add_course(course)
-                    return semester_index
+            if self.semesters[semester_index].course_valid(course):
+                self.semesters[semester_index].add_course(course)
+                return semester_index
         semester = self.add_semester(position=len(self.semesters))
         semester.add_course(course)
         return semester.position
 
 
 class Semester:
-
     is_full = False
     credit_hours = 0
 
-    def __init__(self, max_credits, position):
+    def __init__(self, max_credits, position, past=False):
         self.max_credits = max_credits
         self.position = position
         self.courses = list()
+        self.past = past
+
+    def course_valid(self, course):
+        if not self.is_full and not self.past and self.credit_hours + course.get_credit_hours() <= self.max_credits:
+            if course.season == 2:
+                return self.position % 2 == 0
+            elif course.season == 1:
+                return self.position % 2 == 1
+            else:
+                return True
```

Comment:
Why return True if course.season isn't 1 or 2? Is there another possible value for course.season?

Reply:
Season is only defined when the course can only be taken in a first or second semester, if no such attribute is defined, then it should return true regardless if it is one season or the other.